### PR TITLE
[Doc] project structure topic

### PR DESCRIPTION
**Pull request by jcalcaben**
## Description

This PR revises the project structure and routing tutorial from @rossmc .
Content about the project structure has been expanded on and the routing parts have been moved to the existing routing topic in peregrine. The existing routing topic has also been updated to reflect the current implementation details.

## Related Issue

Closes PWA-300

## Acceptance 

Any developer

### Verification Stakeholders

Any developer

### Verification Steps

1. Navigate to the `pwa-devdocs` directory
1. Run the linter on the modified pages:
    ```sh
    npm run lint src/peregrine/routing/index.md
    npm run lint src/tutorials/pwa-studio-fundamentals/
    ```
1. Verify no linting errors.
1. Build the preview site: `npm run develop`
1. Verify site builds with no errors
1. Navigate to the _Tutorials_ section
1. Verify _Project structure_ link appears on the left nav

## Checklist

* I have updated the documentation accordingly, if necessary.
* I have added tests to cover my changes, if necessary.

### CODE REVIEW

File: pwa-devdocs/src/peregrine/routing/index.md [hunk 1]
```diff
@@ -9,141 +9,76 @@ Every URL request fetches new HTML from the server and the browser loads the ent
 This approach is inefficient because the same assets get loaded every time site navigation occurs.
 
 For a single-page application (SPA), such as a progressive web app, routing is performed on the client side.
-Single-page applications do not reload the browser during internal navigation.
+Single-page applications do not reload the browser for internal links.
```
Comment (zetlen) [inline pwa-devdocs/src/peregrine/routing/index.md]: "Navigation" includes the forward and back buttons as well as form-driven navigations. "Internal links" seems less accurate to me.

File: pwa-devdocs/src/peregrine/routing/index.md [hunk 2]
```diff
@@ -9,141 +9,76 @@ Every URL request fetches new HTML from the server and the browser loads the ent
...
-It also uses the [React Context][] API to create a `Consumer` and `Provider` pair.
-These components allow the router to share information to any of its children without explicitly passing props down the tree.
+### Client-side routing
 
-Since routes can affect different pieces of a page, the MagentoRouter component should be located near the top of the React DOM tree.
+Client-side routing happens inside the storefront application.
+When a user navigates inside the application, it updates the relevant pieces instead of refreshing the entire page to update content.
 
-### Page
+Since, Venia is a single-page application, it uses client-side routing for internal navigation.
```
Comment (zetlen) [inline pwa-devdocs/src/peregrine/routing/index.md]: Extra comma?

File: pwa-devdocs/src/peregrine/routing/index.md [hunk 3]
```diff
@@ -9,141 +9,76 @@ Every URL request fetches new HTML from the server and the browser loads the ent
...
+but it is not the only possible workflow for routing.
 
-**Tip:**
-Use the **Page** and **MagentoRouter** components to integrate the PWA Studio routing feature into your own PWA project.
-{: .bs-callout .bs-callout-tip}
+### Initial request
 
-### MagentoRouteHandler
+Venia's UPWARD server handles the initial request to the storefront application.
+It's `upward.yml` configuration tells the server to return an `index.html` page created from a [template file][].
```
Comment (zetlen) [inline pwa-devdocs/src/peregrine/routing/index.md]:
```suggestion
Its `upward.yml` configuration tells the server to return an `index.html` page created at build time by Webpack, via the [HtmlWebpackPlugin][].
```

File: pwa-devdocs/src/peregrine/routing/index.md [hunk 4]
```diff
@@ -9,141 +9,76 @@ Every URL request fetches new HTML from the server and the browser loads the ent
...
+The content of this file is the same for all page types.
 
-The **MagentoRouteHandler** component uses the route information, passed in by the MagentoRouter `Provider`, to determine which RootComponent to render.
-It does this using the **resolveUnknownRoute** and **fetchRootComponent** helper components.
+After the browser loads the application shell, routing is handled client-side.
```
Comment (zetlen) [inline pwa-devdocs/src/peregrine/routing/index.md]:
```suggestion
After the browser loads the application shell, routing is handled client-side by React Router.
```

File: pwa-devdocs/src/peregrine/routing/index.md [hunk 5]
```diff
@@ -9,141 +9,76 @@ Every URL request fetches new HTML from the server and the browser loads the ent
...
-While the helper components resolves the route and gets the correct root component, the handler renders a loading message component.
-When a root component is successfully loaded it uses the current application state to render the correct content on the page.
-If an error occurs, such as a 404 error, the handler renders an error message component.
+### Routes component
 
-See [Loading and error message components][] to learn how to specify the loading and error message component.
+Used inside the [App component][], the [Routes component][] provides the switch logic for deciding which component to use to render the main content for the current route.
 
-#### resolveUnknownRoute
+Venia has routes not defined in the Magento backend, such as _Create account_ or full page _Checkout_.
```
Comment (zetlen) [inline pwa-devdocs/src/peregrine/routing/index.md]:
```suggestion
Routes in a PWA do not have to correspond to routes in the old Magento storefront. Venia has routes not defined in the Magento backend, such as _Create account_ or full page _Checkout_.
```

File: pwa-devdocs/src/peregrine/routing/index.md [hunk 6]
```diff
@@ -9,141 +9,76 @@ Every URL request fetches new HTML from the server and the browser loads the ent
...
+The [MagentoRoute component][] handles the routes that are Magento-specific, such as a product or category page.
+It uses its [Peregrine talon][] counterpart to determine which component to display.
 
-The `roots-manifest.json` file contains information about the RootComponent chunks generated during a build.
-The file itself is also generated during a build.
+### useMagentoRoute() talon
 
-On subsequent navigations within the PWA, this component sends a request to the Magento backend asking for information about a specific route.
+The [`useMagentoRoute()`][] talon returns the correct component for a page type. It uses the [`getRouteComponent()`][] helper function to get the **root component** associated with a page type.
+It uses the `resolveUnknownRoute()` function to determine the page type for a route and retrieves the root component associated with that type using the global `fetchRootComponent` module.
```
Comment (zetlen) [inline pwa-devdocs/src/peregrine/routing/index.md]:
```suggestion
It uses the `resolveUnknownRoute()` function to determine the page type for a route and retrieves the root component associated with that type using a global `fetchRootComponent` function.
The `fetchRootComponent` function is autogenerated at build time by the Buildpack RootComponentsPlugin, which searches for RootComponents in the project and builds a manifest.
```

File: pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md [hunk 7]
```diff
@@ -0,0 +1,139 @@
...
+
+The ServiceWorker directory contains all the logic for providing [service worker features][].
+Examples of service workers features include push notifications, background sync, and even offline mode for your storefront.
+
+The service worker also gives you the ability to intercept and handle network requests.
+This gives you better control over your site's client-side caching logic.
+
+#### `__tests__` directory
+
+The `__tests__` directory is a suggested directory for your project's test files.
```
Comment (zetlen) [inline pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md]:
```suggestion
The `__tests__` directory is a suggested directory for your project's test files.
Most JavaScript test frameworks find test files using a file path pattern, which by default is usually all `__tests__` directories.
```

File: pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md [hunk 8]
```diff
@@ -0,0 +1,139 @@
...
+This directory comes from the Venia concept project template, which uses [Jest][] for testing.
```
Comment (zetlen) [inline pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md]:
```suggestion
This directory comes from the Venia concept project template, which uses [Jest][] for testing.
Jest recommends that developers make multiple `__tests__` directories nearby the components they are testing.
```

File: pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md [hunk 9]
```diff
@@ -0,0 +1,139 @@
...
+The `status`, `headers`, and `body` values defined in the default `upward.yml` file uses values from the `veniaResponse` object.
+This object is defined in the [venia-ui package's `upward.yml` file][].
+
+Edit this file if you want to change how your project's UPWARD server responds to requests.
+
+#### `template.html`
+
+The `template.html` file contains the template for the server-side rendered HTML code for your storefront's application shell.
+
+During the build process, the build script injects this file with dynamic data, such as project variables and imports for Webpack-generated Javascript bundles.
```
Comment (zetlen) [inline pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md]:
```suggestion
During the build process, the [HtmlWebpackPlugin][] injects this file with dynamic data, such as project variables and imports for Webpack-generated Javascript bundles.
```

File: pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md [hunk 10]
```diff
@@ -0,0 +1,139 @@
...
+[modular components]: <{%link venia-pwa-concept/features/modular-components/index.md %}>
+[project setup]: <{%link tutorials/pwa-studio-fundamentals/index.md %}>
+[`configurewebpack`]: <{%link pwa-buildpack/reference/configure-webpack/index.md %}>
+[upward]: <{%link technologies/upward/index.md %}>
+[load environment file]: <{%link pwa-buildpack/reference/buildpack-cli/load-env/index.md %}#programmatic-api>
+[state management]: <{%link technologies/basic-concepts/state-management/index.md %}>
+
+[service worker features]: https://developers.google.com/web/fundamentals/primers/service-workers
+[jest]: https://jestjs.io/
+[venia-ui package's `upward.yml` file]: https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/upward.yml
```
Comment (zetlen) [inline pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md]: [HtmlWebpackPlugin]: https://webpack.js.org/plugins/html-webpack-plugin/

File: pwa-devdocs/src/peregrine/routing/index.md [hunk 11]
```diff
@@ -9,141 +9,76 @@ Every URL request fetches new HTML from the server and the browser loads the ent
...
-[ErrorView]:t://github.com/magento/pwa-studio/blob/develop/packages/venia-concept/src/components/ErrorView/errorView.js
-[App]: https://github.com/magento/pwa-studio/blob/develop/packages/venia-concept/src/components/App/app.js
-[Loading and error message components]: #loading-and-error-message-components
+[template file]: https://github.com/magento/pwa-studio/blob/develop/packages/venia-concept/template.html
+[app component]: https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/App/app.js
+[routes component]: https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/Routes/routes.js
+[magentoroute component]: https://github.com/magento/pwa-studio/blob/develop/packages/venia-ui/lib/components/MagentoRoute/magentoRoute.js
+[`usemagentoroute()`]: https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/talons/MagentoRoute/useMagentoRoute.js
+[`getroutecomponent()`]: https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/talons/MagentoRoute/getRouteComponent.js
+[`resolveunknownroute()`]: https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/Router/resolveUnknownRoute.js
```
Comment (zetlen) [inline pwa-devdocs/src/peregrine/routing/index.md]:
```suggestion
[`resolveunknownroute()`]: https://github.com/magento/pwa-studio/blob/develop/packages/peregrine/lib/Router/resolveUnknownRoute.js
[HtmlWebpackPlugin]: https://webpack.js.org/plugins/html-webpack-plugin/
```

File: pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md (hunk 10 above)
Comment (jcalcaben) [reply pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md]: is this a blank comment?

Comment (zetlen) [reply pwa-devdocs/src/tutorials/pwa-studio-fundamentals/project-structure/index.md]: What? Oh man. Edited. It was supposed to add a link to HtmlWebpackPlugin

**Comment by dpatil-magento** [comment]
Verification steps pass.

**Pull request merged**